User guide for istioctl managing webhooks (#5162)
istio/istio.io: content/en/docs/setup/install/webhook/index.md

---
title: Istio Webhook Management [Experimental]
description: Shows how to manage webhooks in Istio through istioctl.

> weight: 10

Author: Done.

keywords: [security,webhook]
---

{{< boilerplate experimental-feature-warning >}}

> Istio has two webhooks (Galley and Sidecar Injector). In current Istio implementation,

Reviewer: ```suggestion
Istio has two webhooks: Galley and Sidecar Injector.
```

Author: Done.

> Galley and Sidecar Injector manage their own webhook configurations, which from

Reviewer: ```suggestion
Galley and Sidecar Injector manage their own webhook configurations, which from a
```

Author: Done.

security perspective is not recommended because a compromised webhook may conduct
privilege escalation attacks.

This task shows how to use `istioctl`, instead of Galley and Sidecar Injector, to
manage the webhook configurations of Galley and Sidecar Injector.

## Before you begin

* Create a Kubernetes cluster with Istio installed. In the installation,
[`global.operatorManageWebhooks`]({{< github_file >}}/install/kubernetes/helm/istio/values.yaml) should

> be set as true, and [DNS certificates should be configured](/docs/tasks/security/dns-cert).

Reviewer: ```suggestion
be set to `true`, and [DNS certificates should be configured](/docs/tasks/security/dns-cert).
```

Author: Done.

Istio installation guides can be found [here](/docs/setup/install).

> * Install [`jq`](https://stedolan.github.io/jq/) for json parsing.

Reviewer: ```suggestion
* Install [`jq`](https://stedolan.github.io/jq/) for JSON parsing.
```

Author: Done.

## Check webhook certificates

> The webhook certificates generated are stored in the secrets specified in the configuration.

Reviewer: What is the configuration referred to here? Add links or be more specific.
I'd just remove this sentence.

Author: Done: removed this sentence.

> 1.  Check that webhook certificates have been generated:

Reviewer: No need to use a numbered list for only one item.
```suggestion
To display the DNS names in the webhook certificates of Galley and Sidecar Injector, run the following commands:
```

Author: Done.

    {{< text bash >}}
    $ kubectl get secret dns.istio-galley-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
    $ kubectl get secret dns.istio-sidecar-injector-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
    {{< /text >}}

    The output from the above commands should include the DNS names of Galley and Sidecar Injector, respectively:

    {{< text plain >}}
    X509v3 Subject Alternative Name:
      DNS:istio-galley.istio-system.svc, DNS:istio-galley.istio-system
    {{< /text >}}

    {{< text plain >}}
    X509v3 Subject Alternative Name:
      DNS:istio-sidecar-injector.istio-system.svc, DNS:istio-sidecar-injector.istio-system
    {{< /text >}}

## `istioctl` enables webhook configurations

1.  Generate `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` by running the following
command:

    {{< text bash >}}
    $ helm template \
        --name=istio \
        --namespace=istio-system \

>         --values install/kubernetes/helm/istio/example-values/values-istio-dns-cert.yaml \

Reviewer: Should these values be customized? In which cases? How?

Author: Done: explanations added.

    {{< /text >}}

<!-- TODO (lei-tang): improve the UX for obtain MutatingWebhookConfiguration -->
1.  From `istio-webhook-config.yaml`, search `'kind: MutatingWebhookConfiguration'` and save
the `MutatingWebhookConfiguration` of Sidecar Injector to `sidecar-injector-webhook.yaml`. The following

> is a `MutatingWebhookConfiguration` in an example `istio-webhook-config.yaml`.

Reviewer 2: Why not just copy the following two blocks of yaml configs instead?

Author: Because a user may want to customize the namespace, etc in the yaml config, the two blocks of yaml configs are not copied.

Reviewer 2: But as far as I see, you didn't tell them to how to customize, the `helm template` command generates a yaml file that contains the config which doesn't seem to be different from the two blocks you gave in the following. So if you want users to customize, why not explicitly mentioning it in the following? We were doing this for other tasks.

Author: Added an explanation that the command in the guide uses the default Istio configuration plus the DNS certificate configuration in `values-istio-dns-cert.yaml`, as well as a link to the helm template document. This guide does not require users to customize the namespace, and etc, but the namespace and etc in the validatingwebhookconfiguration and mutatingwebhookconfiguration should be consistent with the users' Istio installation. Therefore, the guide does not provide two hard-coded blocks of yaml configs of validatingwebhookconfiguration and mutatingwebhookconfiguration to copy and paste.

Reviewer 3: Please ensure that the actions the reader needs to take are clear. I was unable to discern all actions needed here.

is a `ValidatingWebhookConfiguration` in an example `istio-webhook-config.yaml`.

    {{< text yaml >}}
    apiVersion: admissionregistration.k8s.io/v1beta1
    kind: ValidatingWebhookConfiguration
    metadata:
      name: istio-galley
      labels:
        app: galley
        chart: galley
        heritage: Tiller
        release: istio
        istio: galley
    webhooks:
      - name: pilot.validation.istio.io
        clientConfig:
          service:
            name: istio-galley
            namespace: istio-system
            path: "/admitpilot"
          caBundle: ""
        rules:
          - operations:
            - CREATE
            - UPDATE
            apiGroups:
            - config.istio.io
            ... SKIPPED
        failurePolicy: Fail
        sideEffects: None
    {{< /text >}}

1.  Enable webhook configurations through `istioctl`:

    {{< text bash >}}
    $ istioctl experimental post-install webhook enable --validation --webhook-secret dns.istio-galley-service-account \
        --namespace istio-system --validation-path galley-webhook.yaml \
        --injection-path sidecar-injector-webhook.yaml
    {{< /text >}}

> 1.  Check the Sidecar Injector webhook is working:

Reviewer: Explain what this test is doing.

Author: Done: explanations added.

    {{< text bash >}}
    $ kubectl create namespace test-injection; kubectl label namespaces test-injection istio-injection=enabled
    $ kubectl run --generator=run-pod/v1 --image=nginx nginx-app --port=80 -n test-injection
    $ kubectl get pod -n test-injection

>     {{< /text >}}

Reviewer: Delete the `test-injection` namespace!

Author: The deleting of test namespaces is part of the cleanup in the guide: "you may delete the testing cluster created at the beginning of this tutorial."

    The output from the `get pod` command should show the following output (`2/2` means that

>     the Sidecar Injector webhook injects a sidecar to the example pod):

Reviewer: ```suggestion
    the Sidecar Injector webhook injected a sidecar into the example pod):
```

Author: Done.

    {{< text plain >}}
    NAME    READY   STATUS    RESTARTS   AGE
    nginx-app   2/2     Running   0          10s
    {{< /text >}}

> 1.  Check the validation webhook is working:

Reviewer: ```suggestion
1.  Check that the validation webhook is working:
```

Author: Done.

    {{< text bash >}}
    $ cat <<EOF > ./invalid-gateway.yaml
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: invalid-gateway
    spec:
      selector:
        # DO NOT CHANGE THESE LABELS
        # The ingressgateway is defined in install/kubernetes/helm/istio/values.yaml
        # with these labels
        istio: ingressgateway
    EOF
    $ kubectl create namespace test-validation

>     $ kubectl apply -f invalid-gateway.yaml -n test-validation

Reviewer: Simplify:
```
$ kubectl create namespace test-validation
$ kubectl apply -n test-validation -f - <<EOF
... <your yaml above>
EOF
```

Author: Done.

>     {{< /text >}}

Reviewer: Delete the `test-validation` namespace!

Author: The deleting of test namespaces is part of the cleanup in the guide: "you may delete the testing cluster created at the beginning of this tutorial."

    The output from the gateway creation command should show the following output (the error

>     in the output indicates that the validation webhook checked the gateway yaml file):

Reviewer: ```suggestion
    in the output indicates that the validation webhook checked the gateway YAML file):
```

Author: Done.

    {{< /text >}}

1.  Show the configuration of Galley with the webhook configuration name being `istio-galley`:

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --injection=false --validation-config=istio-galley
    {{< /text >}}

## `istioctl` disables webhook configurations

1.  Disable the configurations of Galley and Sidecar Injector with their default webhook configuration names:

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable
    {{< /text >}}

1.  Disable the configuration of Sidecar Injector with the webhook configuration name being `istio-sidecar-injector`:

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

1.  Disable the configuration of Galley with the webhook configuration name being `istio-galley`:

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --injection=false --validation-config=istio-galley
    {{< /text >}}

## Cleanup

After completing this tutorial, you may delete the testing cluster created

> at the beginning of this tutorial.

Reviewer 2: This is not meaningful information. Can you refer to other tasks to see if you can improve?

Author: To safely test an experimental feature, this guide creates a test cluster and the cleanup simply deletes the test cluster. If you want to add individual commands to the guide to revert the commands executed in the guide, please let me know and I can add them.

Reviewer 2: Yes, is there a clean reverting approach?

Author: Added the command to delete the resources created.

---
title: Istio Webhook Management [Experimental]
description: Shows how to manage webhooks in Istio through istioctl.
weight: 100
keywords: [security,webhook]
---

{{< boilerplate experimental-feature-warning >}}

> Istio has two webhooks: Galley and Sidecar Injector. By default,

Author: Done.

> Galley and Sidecar Injector manage their own webhook configurations, which from the

Author: Done.

> security perspective is not recommended because a compromised webhook may conduct

Author: Done.

privilege escalation attacks.

> This task shows how to use `istioctl`, instead of Galley and Sidecar Injector, to

Author: Done.

> manage the webhook configurations of Galley and Sidecar Injector.

Author: Done.

## Before you begin

* Create a Kubernetes cluster with Istio installed. In the installation,
[`global.operatorManageWebhooks`]({{< github_file >}}/install/kubernetes/helm/istio/values.yaml) should
be set to `true`, and [DNS certificates should be configured](/docs/tasks/security/dns-cert).
Istio installation guides can be found [here](/docs/setup/install).

* Install [`jq`](https://stedolan.github.io/jq/) for JSON parsing.

## Check webhook certificates

> To display the DNS names in the webhook certificates of Galley and Sidecar Injector, run the following commands:

Author: Done.

{{< text bash >}}
$ kubectl get secret dns.istio-galley-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
$ kubectl get secret dns.istio-sidecar-injector-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
{{< /text >}}

> The output from the above commands should include the DNS names of Galley and Sidecar Injector, respectively:

Author: Done.

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-galley.istio-system.svc, DNS:istio-galley.istio-system
{{< /text >}}

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-sidecar-injector.istio-system.svc, DNS:istio-sidecar-injector.istio-system
{{< /text >}}

## `istioctl` enables webhook configurations

> 1.  Generate `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` by running the following

Reviewer 3: ```suggestion
1.  To generate the `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` configuration files, run the following
```

Author: Done.

> command. The YAML file [`values-istio-dns-cert.yaml`]({{< github_file >}}/install/kubernetes/helm/istio/example-values/values-istio-dns-cert.yaml)

Reviewer 3: ```suggestion
command. The [`values-istio-dns-cert.yaml`]({{< github_file >}}/install/kubernetes/helm/istio/example-values/values-istio-dns-cert.yaml) YAML file
```

Author: Done.

> contains an example DNS certificate configuration (details in [the certificate guide](/docs/tasks/security/dns-cert)).

Reviewer 3: ```suggestion
contains an example DNS certificate configuration. Learn more in [the certificate guide](/docs/tasks/security/dns-cert).
```

Author: Done.

> The following command uses the default Istio configuration plus the DNS certificate configuration in `values-istio-dns-cert.yaml`.

Reviewer 3: ```suggestion
The following command uses the following configurations:

    - The default Istio configuration
    - The DNS certificate configuration in `values-istio-dns-cert.yaml`
```

Author: Done.

> The document of the helm template command can be found in the [link](https://helm.sh/docs/helm/#helm-template).

Reviewer 3: ```suggestion
You can find more information in the `helm template` command [documentation](https://helm.sh/docs/helm/#helm-template).
```

Author: Done.

1.  From `istio-webhook-config.yaml`, search `'kind: MutatingWebhookConfiguration'` and save
the `MutatingWebhookConfiguration` of Sidecar Injector to `sidecar-injector-webhook.yaml`. The following
is a `MutatingWebhookConfiguration` in an example `istio-webhook-config.yaml`.

    {{< text yaml >}}
    apiVersion: admissionregistration.k8s.io/v1beta1
    kind: MutatingWebhookConfiguration
    metadata:
      name: istio-sidecar-injector
      labels:
        app: sidecarInjectorWebhook
        chart: sidecarInjectorWebhook
        heritage: Tiller
        release: istio
    webhooks:
      - name: sidecar-injector.istio.io
        clientConfig:
          service:
            name: istio-sidecar-injector
            namespace: istio-system
            path: "/inject"
          caBundle: ""
        rules:
          - operations: [ "CREATE" ]
            apiGroups: [""]
            apiVersions: ["v1"]
            resources: ["pods"]
        failurePolicy: Fail
        namespaceSelector:
          matchLabels:
            istio-injection: enabled
    {{< /text >}}

<!-- TODO (lei-tang): improve the UX for obtain ValidatingWebhookConfiguration -->

> 1.  From `istio-webhook-config.yaml`, search `'kind: ValidatingWebhookConfiguration'` and save

Reviewer 3: The actions the reader needs to take are not clear. How exactly can the reader determine what to copy and save in the new file? Where should that file be stored?

Author: Done.

the `ValidatingWebhookConfiguration` of Galley to `galley-webhook.yaml`. The following
is a `ValidatingWebhookConfiguration` in an example `istio-webhook-config.yaml` (only
a part of the configuration is shown to save space).

    {{< text yaml >}}
    apiVersion: admissionregistration.k8s.io/v1beta1
    kind: ValidatingWebhookConfiguration
    metadata:
      name: istio-galley
      labels:
        app: galley
        chart: galley
        heritage: Tiller
        release: istio
        istio: galley
    webhooks:
      - name: pilot.validation.istio.io
        clientConfig:
          service:
            name: istio-galley
            namespace: istio-system
            path: "/admitpilot"
          caBundle: ""
        rules:
          - operations:
            - CREATE
            - UPDATE
            apiGroups:
            - config.istio.io
            ... SKIPPED
        failurePolicy: Fail
        sideEffects: None
    {{< /text >}}

> 1.  Enable webhook configurations through `istioctl`:

Reviewer 3: ```suggestion
1. Use `istioctl` to enable the webhook configurations:
```

Author: Done.

    {{< text bash >}}
    $ istioctl experimental post-install webhook enable --validation --webhook-secret dns.istio-galley-service-account \
        --namespace istio-system --validation-path galley-webhook.yaml \
        --injection-path sidecar-injector-webhook.yaml
    {{< /text >}}

> 1.  Check the Sidecar Injector webhook is working by verifying that Sidecar Injector injects a

Reviewer 3: ```suggestion
1.  To check that the sidecar injector webhook is working, verify that the webhook injects a
```

Author: Done.

> sidecar container into an example pod:

Reviewer 3: ```suggestion
sidecar container into an example pod with the following commands:
```

Author: Done.

    {{< text bash >}}
    $ kubectl create namespace test-injection; kubectl label namespaces test-injection istio-injection=enabled
    $ kubectl run --generator=run-pod/v1 --image=nginx nginx-app --port=80 -n test-injection
    $ kubectl get pod -n test-injection
    {{< /text >}}

>     The output from the `get pod` command should show the following output (`2/2` means that

Reviewer 3: ```suggestion
    The output from the `get pod` command should show the following output. The `2/2` value means that
```

Author: Done.

>     the Sidecar Injector webhook injected a sidecar into the example pod):

Reviewer 3: ```suggestion
    the webhook injected a sidecar into the example pod:
```

Author: Done.

    {{< text plain >}}
    NAME    READY   STATUS    RESTARTS   AGE
    nginx-app   2/2     Running   0          10s
    {{< /text >}}

1.  Check that the validation webhook is working:

    {{< text bash >}}
    $ kubectl create namespace test-validation
    $ kubectl apply -n test-validation -f - <<EOF
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: invalid-gateway
    spec:
      selector:
        # DO NOT CHANGE THESE LABELS
        # The ingressgateway is defined in install/kubernetes/helm/istio/values.yaml
        # with these labels
        istio: ingressgateway
    EOF
    {{< /text >}}

>     The output from the gateway creation command should show the following output (the error

Reviewer 3: ```suggestion
    The output from the gateway creation command should show the following output. The error
```

Author: Done.

>     in the output indicates that the validation webhook checked the gateway YAML file):

Reviewer 3: ```suggestion
    in the output indicates that the validation webhook checked the gateway's configuration YAML file:
```

Author: Done.

    {{< text plain >}}
    Error from server: error when creating "invalid-gateway.yaml": admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: gateway must have at least one server
    {{< /text >}}

## Show webhook configurations

> 1.  Show the configurations of Galley and Sidecar Injector with their default webhook configuration names:

Reviewer 3: ```suggestion
1.  Show the configurations of Galley and the sidecar injector with their default configuration names:
```

Author: Done.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status
    {{< /text >}}

> 1.  Show the configuration of Sidecar Injector with the webhook configuration name being `istio-sidecar-injector`:

Reviewer 3: ```suggestion
1.  Show the configuration of the sidecar injector if the webhook's configuration name is `istio-sidecar-injector`:
```

Author: Done.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

> 1.  Show the configuration of Galley with the webhook configuration name being `istio-galley`:

Reviewer 3: ```suggestion
1.  Show the configuration of Galley if the webhook's configuration name is `istio-galley`:
```

Author: Done.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --injection=false --validation-config=istio-galley
    {{< /text >}}

## Disable webhook configurations

> 1.  Disable the configurations of Galley and Sidecar Injector with their default webhook configuration names:

Reviewer 3: ```suggestion
1.  Disable Galley and the sidecar injector when using their default webhook configuration names:
```

Author: Done.

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable
    {{< /text >}}

> 1.  Disable the configuration of Sidecar Injector with the webhook configuration name being `istio-sidecar-injector`:

Reviewer 3: ```suggestion
1.  Disable the sidecar injector if the webhook's configuration name is `istio-sidecar-injector`:
```

Author: Done.

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

> 1.  Disable the configuration of Galley with the webhook configuration name being `istio-galley`:

Reviewer 3: ```suggestion
1.  Disable the configuration of Galley if the webhook's configuration name is `istio-galley`:
```

Author: Done.

---
title: Istio Webhook Management [Experimental]
description: Shows how to manage webhooks in Istio through istioctl.
weight: 100
keywords: [security,webhook]
---

{{< boilerplate experimental-feature-warning >}}

Istio has two webhooks: Galley and the sidecar injector. By default,
these webhooks manage their own configurations. From a
security perspective, this default behavior is not recommended because a compromised webhook could then conduct
privilege escalation attacks.

This task shows how to use `istioctl`, instead of the webhooks, to
manage their configurations.

## Before you begin

> * Create a Kubernetes cluster with Istio installed. In the installation,

Reviewer 3: What installation guide do readers need to follow? What configuration profile should they use? Please assume folks have deployed Istio using one of our existing configuration profiles and provide them here with the steps they need to take to modify said deployment before starting the task.

Author: Revised as follows (the installation guide reuses the one in the /docs/tasks/security/dns-cert):

Install Istio with [DNS certificates configured](/docs/tasks/security/dns-cert) and
[`global.operatorManageWebhooks`]({{< github_file >}}/install/kubernetes/helm/istio/values.yaml) set to `true`.

[`global.operatorManageWebhooks`]({{< github_file >}}/install/kubernetes/helm/istio/values.yaml) should
be set to `true`, and [DNS certificates should be configured](/docs/tasks/security/dns-cert).
Istio installation guides can be found [here](/docs/setup/install).

* Install [`jq`](https://stedolan.github.io/jq/) for JSON parsing.

## Check webhook certificates

> To display the DNS names in the webhook certificates of Galley and the sidecar injector, run the following commands:

Reviewer 3: Do one first and then the other. Ask readers display the DNS names for the Galley webhook then check the output. Then, ask readers to do the same for the sidecar injector. Remember to number all four steps and split the commands.
For readers, both the text and the commands appear way more complicated than they are.
Avoid using `|` in commands for that same reason.
Be accurate in your description of what the commands do. Not every reader will understand or be as familiar as you are with the commands and tools you are using.
```suggestion
To display the DNS names in the webhook certificates of Galley and the sidecar injector, we need to get the secret from Kubernetes, parse it, decode it, and apply the key to see the text output with the following  commands:
```

Author: Revised accordingly.

    the webhook injected a sidecar into the example pod:

    {{< text plain >}}
    NAME    READY   STATUS    RESTARTS   AGE
    nginx-app   2/2     Running   0          10s
    {{< /text >}}

1.  Check that the validation webhook is working:

    {{< text bash >}}
    $ kubectl create namespace test-validation
    $ kubectl apply -n test-validation -f - <<EOF
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: invalid-gateway
    spec:
      selector:
        # DO NOT CHANGE THESE LABELS
        # The ingressgateway is defined in install/kubernetes/helm/istio/values.yaml
        # with these labels
        istio: ingressgateway
    EOF
    {{< /text >}}

    The output from the gateway creation command should show the following output. The error
    in the output indicates that the validation webhook checked the gateway's configuration YAML file:

    {{< text plain >}}
    Error from server: error when creating "invalid-gateway.yaml": admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: gateway must have at least one server
    {{< /text >}}

## Show webhook configurations

> 1.  Show the configurations of Galley and the sidecar injector with their default configuration names:

Reviewer 3: ```suggestion
1.  Show the configuration of the webhooks:
```

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status
    {{< /text >}}

> 1.  Show the configuration of the sidecar injector if the webhook's configuration name is `istio-sidecar-injector`:

Reviewer 3: ```suggestion
1.  If you named the sidecar injector's configuration `istio-sidecar-injector`, use the following command to show the configuration of the sidecar injector:
```

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

> 1.  Show the configuration of Galley if the webhook's configuration name is `istio-galley`:

Reviewer 3: ```suggestion
1.  If you named Galley's configuration `istio-galley`, show Galley's configuration with the following command:
```

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --injection=false --validation-config=istio-galley
    {{< /text >}}

## Disable webhook configurations

1.  Disable Galley and the sidecar injector when using their default webhook configuration names:

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable
    {{< /text >}}

> 1.  Disable the sidecar injector if the webhook's configuration name is `istio-sidecar-injector`:

Reviewer 3: ```suggestion
1.  If you named the sidecar injector's configuration `istio-sidecar-injector`, disable the webhook with the following command:
```

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

> 1.  Disable the configuration of Galley if the webhook's configuration name is `istio-galley`:

Reviewer 3: ```suggestion
1.   If you name Galleys's configuration `istio-galley`, disable the webhook with the following command:
```

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --injection=false --validation-config=istio-galley
    {{< /text >}}

## Cleanup

> After completing this tutorial, you may delete the testing cluster created

Reviewer 3: Please do not make a test cluster a requirement for tasks. Instead, make sure features can be enabled and disabled on existing deployments.

Author: Creating a test cluster is to safely try this feature without affecting existing deployments.
This feature requires generating DNS certificates and disabling Galley and the sidecar injector managing their own webhooks. Thus currently this feature can only be enabled and disabled on a new deployment.
